File: adventofcode/2024/04_1.cpp
#include <bits/stdc++.h>
#define int long long

using namespace std;
vector<string> grid;

int rows, cols;
vector<vector<pair<int, int>>> dirs = {
    {{0, 0}, {0, 1}, {0, 2}, {0, 3}},
    {{0, 0}, {0, -1}, {0, -2}, {0, -3}},
    {{0, 0}, {1, 0}, {2, 0}, {3, 0}},
    {{0, 0}, {-1, 0}, {-2, 0}, {-3, 0}},
    {{0, 0}, {1, 1}, {2, 2}, {3, 3}},
    {{0, 0}, {-1, -1}, {-2, -2}, {-3, -3}},
    {{0, 0}, {1, -1}, {2, -2}, {3, -3}},
    {{0, 0}, {-1, 1}, {-2, 2}, {-3, 3}},
};
const string XMAS = "XMAS";

int search(int i, int j) {
    int ans = 0;
    for(auto d: dirs) {
        bool found = true;
        for(int k=0; k<4; k++) {
            auto [dx, dy] = d[k];
            int x=i+dx;
            int y=j+dy;
            if (x<0 || x>=rows || y<0 || y>=cols || grid[x][y] != XMAS[k]) {
                found = false; break;
            }
        }
        if (found) ans++;
    }
    return ans;
}

signed main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    #ifndef ONLINE_JUDGE 
        freopen("../input.txt", "r", stdin); 
        freopen("../output.txt", "w", stdout); 
    #endif
    const int LINES=140;

    for(int i=0; i<LINES; i++)
    {
        string s; getline(cin, s);
        grid.push_back(s);
    }
    rows = grid.size();
    cols = grid[0].size();
    int ans = 0;
    for(int i=0; i<rows; i++) {
        for(int j=0; j<cols; j++) {
            ans += search(i, j);
        }
    }
    cout << ans;
    return 0;
}

File: adventofcode/2024/python/04_1.py
import sys

dirs = [
    [(0, 0), (0, 1), (0, 2), (0, 3)],  # Horizontal right
    [(0, 0), (0, -1), (0, -2), (0, -3)],  # Horizontal left
    [(0, 0), (1, 0), (2, 0), (3, 0)],  # Vertical down
    [(0, 0), (-1, 0), (-2, 0), (-3, 0)],  # Vertical up
    [(0, 0), (1, 1), (2, 2), (3, 3)],  # Diagonal down-right
    [(0, 0), (-1, -1), (-2, -2), (-3, -3)],  # Diagonal up-left
    [(0, 0), (1, -1), (2, -2), (3, -3)],  # Diagonal down-left
    [(0, 0), (-1, 1), (-2, 2), (-3, 3)],  # Diagonal up-right
]

XMAS = "XMAS"

def search(grid, i, j):
    rows, cols = len(grid), len(grid[0])
    ans = 0
    for d in dirs:
        found = True
        for k in range(4):
            dx, dy = d[k]
            x, y = i+dx, j+dy
            if x<0 or x>=rows or y<0 or y>=cols or grid[x][y] != XMAS[k]:
                found = False
                break
        if found:
            ans += 1
    return ans

def main():
    sys.stdin = open("../input.txt", "r")
    sys.stdout = open("../output.txt", "w")
    
    LINES = 140
    grid = [input().strip() for _ in range(LINES)]
    rows, cols = len(grid), len(grid[0])
    ans = 0;
    
    for i in range(rows):
        for j in range(cols):
            ans += search(grid, i, j)
    
    print(ans)

if __name__ == "__main__":
    main()